### ISSUE TITLE
Grouped Time Series Validation

### ISSUE THREAD

**Issue by Jhixx24** [mljar/mljar-supervised#315]
```python
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from supervised.automl import AutoML # mljar-supervised

def correlation(predictions, targets):
    ranked_preds = predictions.rank(pct=True, method="first")
    return np.corrcoef(ranked_preds, targets)[0, 1]

def score_2(df):
    return correlation(df[PREDICTION_NAME], df[TARGET_NAME])

TARGET_NAME = f"target"
PREDICTION_NAME = f"prediction"

training_data = pd.read_csv("https://numerai-public-datasets.s3-us-west-2.amazonaws.com/latest_numerai_training_data.csv.xz")
tournament_data = pd.read_csv("https://numerai-public-datasets.s3-us-west-2.amazonaws.com/latest_numerai_tournament_data.csv.xz")
training_data = training_data.drop('Unnamed: 0',axis=1)
tournament_data = tournament_data.drop('Unnamed: 0',axis=1)
training_data2 =training_data[0:int(len(training_data)*0.05)]
tournament_data2 =tournament_data[0:int(len(tournament_data)*0.05)]
training_data2
tournament_data2
#############################################################################################
eras = training_data["era"]
#############################################################################################
features = [f for f in training_data2.columns if f.startswith("feature_intelligence1")]
X = training_data[features]
X2 = tournament_data[features]
y = training_data[TARGET_NAME]

#{`Explain`, `Perform`, `Compete`}
automl = AutoML(mode="Explain",algorithms=["Xgboost"],validation_strategy=key)
automl.fit(X, y)

predictions = automl.predict(X2)

pd.Series(predictions)
tournament_data2[PREDICTION_NAME] = predictions
tournament_data2

validation_data = training_data2[training_data2.data_type == "validation"]
validation_correlations = validation_data.groupby("era").apply(score_2)
validation_correlations

#############################################################################################
class TimeSeriesSplitGroups(_BaseKFold):
    def __init__(self, n_splits=5):
        super().__init__(n_splits, shuffle=False, random_state=None)

    def split(self, X, y=None, groups=None):
        X, y, groups = indexable(X, y, groups)
        n_samples = _num_samples(X)
        n_splits = self.n_splits
        n_folds = n_splits + 1
        group_list = np.unique(groups)
        n_groups = len(group_list)
        if n_folds > n_groups:
            raise ValueError(
                ("Cannot have number of folds ={0} greater"
                 " than the number of samples: {1}.").format(n_folds,
                                                             n_groups))
        indices = np.arange(n_samples)
        test_size = (n_groups // n_folds)
        test_starts = range(test_size + n_groups % n_folds,
                            n_groups, test_size)
        test_starts = list(test_starts)[::-1]
        for test_start in test_starts:

            yield (indices[groups.isin(group_list[:test_start])],
                   indices[groups.isin(group_list[test_start:test_start + test_size])])

```

**Comment by pplonski** [comment]
Hi @Jhixx24! I know Numerai data quite well.

This type of cross-validation request can be added, but hard to say when. Would you like to follow an example from Numerai's forum or you have some unique idea for feature engineering that will work with this type of validation?

Going back to the data itself. Please run it on a 10-fold CV and it will work pretty well (better than the example model provided by Numerai as far as I remember). But you need to run on a decent machine for at least 12 hours. Below is the example code how to use MLJAR AutoML:

```python
train = pd.read_csv("numerai_training_data.csv")
x_cols = [f for f in train.columns if "feature" in f]
y_col = "target"

automl = AutoML(
    ml_task="regression",
    mode="Compete",
    total_time_limit=12 * 60 * 60,
)
automl.fit(train[x_cols], train[y_col])
```

**Comment by Jhixx24** [comment]
hi..I am trying to run AutoML but i cannot get it to split the data according to the different "eras" or periods in the dataset. i have tried 5fold cv but it splits and shuffles them without respecting the era. i have added TimeSeriesSplitGroups class from a different script that gets the appropriate eras data but do not know how to implement it in your code. any assistance would be appreciated.

**Comment by Jhixx24** [comment]
so the current implementation of cv will work ?..because i suspect it will not cut the data at the right places.

**Comment by pplonski** [comment]
OK, got it. But my point is, that you don't need to respect "eras" when doing validation. You can simply shuffle samples from different eras and train AutoML with 5-fold or 10-fold CV. I'm using such approach - MLJAR AutoML is part of my ensemble. My performance metric is below:

![image](https://user-images.githubusercontent.com/6959032/108182700-98f88400-7109-11eb-94fb-72b0a8f79428.png)

I'm not using validation data for the training. The last tip from me, I'm using feature neutralization.

**Comment by Jhixx24** [comment]
ahh ok thanks..last question..might you have an idea on how to get rid of this error while installing Auto ML ? RuntimeError: Building llvmlite requires LLVM 10.0.x or 9.0.x, got '11.0.1'. Be sure to set LLVM_CONFIG to the right executable path.

**Comment by pplonski** [comment]
I need more information:
- python version
- system version
- how do you install it?

**Comment by Jhixx24** [comment]
maragaxxx@instance-1:~$ pip3 install mljar-supervised
Requirement already satisfied: mljar-supervised in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages/mljar_supervised-0.8.9-py3.9.egg (0.8.9)
Requirement already satisfied: numpy>=1.18.5 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from mljar-supervised) (1.20.1)
Requirement already satisfied: pandas==1.2.0 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from mljar-supervised) (1.2.0)
Requirement already satisfied: scipy==1.6.0 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from mljar-supervised) (1.6.0)
Requirement already satisfied: scikit-learn==0.23.2 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from mljar-supervised) (0.23.2)
Requirement already satisfied: xgboost==1.2.0 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from mljar-supervised) (1.2.0)
Requirement already satisfied: lightgbm==3.0.0 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from mljar-supervised) (3.0.0)
Requirement already satisfied: catboost==0.24.4 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from mljar-supervised) (0.24.4)
Requirement already satisfied: joblib==0.14.1 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from mljar-supervised) (0.14.1)
Requirement already satisfied: cloudpickle==1.3.0 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from mljar-supervised) (1.3.0)
Collecting pyarrow>=2.0.0
  Using cached pyarrow-3.0.0-cp39-cp39-manylinux2014_x86_64.whl (20.7 MB)
Requirement already satisfied: tabulate==0.8.7 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from mljar-supervised) (0.8.7)
Requirement already satisfied: matplotlib>=3.2.2 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from mljar-supervised) (3.3.4)
Requirement already satisfied: dtreeviz==1.0 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from mljar-supervised) (1.0)
Collecting shap==0.36.0
  Using cached shap-0.36.0-cp39-cp39-linux_x86_64.whl
Requirement already satisfied: seaborn==0.10.1 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from mljar-supervised) (0.10.1)
Collecting wordcloud==1.7.0
  Using cached wordcloud-1.7.0.tar.gz (216 kB)
Requirement already satisfied: category_encoders==2.2.2 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages/category_encoders-2.2.2-py3.9.egg (from mljar-supe
rvised) (2.2.2)
Requirement already satisfied: six in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from catboost==0.24.4->mljar-supervised) (1.15.0)
Requirement already satisfied: graphviz in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from catboost==0.24.4->mljar-supervised) (0.16)
Requirement already satisfied: plotly in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from catboost==0.24.4->mljar-supervised) (4.14.3)
Requirement already satisfied: patsy>=0.5.1 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from category_encoders==2.2.2->mljar-supervised) (0.5.1)
Requirement already satisfied: statsmodels>=0.9.0 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from category_encoders==2.2.2->mljar-supervised) (0.12.
2)
Requirement already satisfied: pytest in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from dtreeviz==1.0->mljar-supervised) (6.2.2)
Requirement already satisfied: colour in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from dtreeviz==1.0->mljar-supervised) (0.1.5)
Requirement already satisfied: pytz>=2017.3 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from pandas==1.2.0->mljar-supervised) (2021.1)
Requirement already satisfied: python-dateutil>=2.7.3 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from pandas==1.2.0->mljar-supervised) (2.8.1)
Requirement already satisfied: threadpoolctl>=2.0.0 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from scikit-learn==0.23.2->mljar-supervised) (2.1.0)
Collecting numba
  Using cached numba-0.51.2-cp39-cp39-linux_x86_64.whl
Collecting slicer
  Using cached slicer-0.0.7-py3-none-any.whl (14 kB)
Collecting tqdm>4.25.0
  Using cached tqdm-4.56.2-py2.py3-none-any.whl (72 kB)
Requirement already satisfied: pillow in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from wordcloud==1.7.0->mljar-supervised) (8.1.0)
Requirement already satisfied: kiwisolver>=1.0.1 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from matplotlib>=3.2.2->mljar-supervised) (1.3.1)
Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.3 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from matplotlib>=3.2.2->mljar-sup
ervised) (2.4.7)

**Comment by Jhixx24** [comment]
'"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/
pip-record-3e8k4yo5/install-record.txt --single-version-externally-managed --compile --install-headers /home/linuxbrew/.linuxbrew/include/python3.9/llvmlite
         cwd: /tmp/pip-install-h6mei8uq/llvmlite_8ae8ea1bba544bd490f89481bd7d3652/
    Complete output (29 lines):
    running install
    running build
    got version from file /tmp/pip-install-h6mei8uq/llvmlite_8ae8ea1bba544bd490f89481bd7d3652/llvmlite/_version.py {'version': '0.34.0', 'full': 'c5889c9e98c6b19d5d
85ebdd982d64a03931f8e2'}
    running build_ext
    /home/linuxbrew/.linuxbrew/opt/python@3.9/bin/python3.9 /tmp/pip-install-h6mei8uq/llvmlite_8ae8ea1bba544bd490f89481bd7d3652/ffi/build.py
    LLVM version... Traceback (most recent call last):
      File "/tmp/pip-install-h6mei8uq/llvmlite_8ae8ea1bba544bd490f89481bd7d3652/ffi/build.py", line 105, in main_posix
        out = subprocess.check_output([llvm_config, '--version'])
      File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/subprocess.py", line 420, in check_output
        return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
      File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/subprocess.py", line 501, in run
        with Popen(*popenargs, **kwargs) as process:
      File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/subprocess.py", line 947, in __init__
        self._execute_child(args, executable, preexec_fn, close_fds,
      File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/subprocess.py", line 1819, in _execute_child
        raise child_exception_type(errno_num, err_msg, err_filename)
    FileNotFoundError: [Errno 2] No such file or directory: 'llvm-config'
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "/tmp/pip-install-h6mei8uq/llvmlite_8ae8ea1bba544bd490f89481bd7d3652/ffi/build.py", line 191, in <module>
        main()
      File "/tmp/pip-install-h6mei8uq/llvmlite_8ae8ea1bba544bd490f89481bd7d3652/ffi/build.py", line 181, in main
        main_posix('linux', '.so')
      File "/tmp/pip-install-h6mei8uq/llvmlite_8ae8ea1bba544bd490f89481bd7d3652/ffi/build.py", line 107, in main_posix
        raise RuntimeError("%s failed executing, please point LLVM_CONFIG "
    RuntimeError: llvm-config failed executing, please point LLVM_CONFIG to the path for llvm-config
    error: command '/home/linuxbrew/.linuxbrew/opt/python@3.9/bin/python3.9' failed with exit code 1
    ----------------------------------------
  Rolling back uninstall of llvmlite
  Moving to /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages/llvmlite-0.36.0rc1.dist-info/
   from /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages/~lvmlite-0.36.0rc1.dist-info
  Moving to /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages/llvmlite/
   from /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages/~lvmlite
ERROR: Command errored out with exit status 1: /home/linuxbrew/.linuxbrew/opt/python@3.9/bin/python3.9 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/
tmp/pip-install-h6mei8uq/llvmlite_8ae8ea1bba544bd490f89481bd7d3652/setup.py'"'"'; __file__='"'"'/tmp/pip-install-h6mei8uq/llvmlite_8ae8ea1bba544bd490f89481bd7d3652/
setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'ex
ec'"'"'))' install --record /tmp/pip-record-3e8k4yo5/install-record.txt --single-version-externally-managed --compile --install-headers /home/linuxbrew/.linuxbrew/i
nclude/python3.9/llvmlite Check the logs for full command output.

**Comment by pplonski** [comment]
Python 3.9 is not yet supported. Please try with Python 3.7

**Comment by Jhixx24** [comment]
Ok..i will try it. Thank you.

**Comment by Jhixx24** [comment]
it worked ! 👍

**Comment by pplonski** [comment]
@Jhixx24 Grouped Time Series validation can be applied with custom validation. Closing the issue. Fixed in #380.
